File: Breadth-First Search/Binary tree/LET_no623_Add One Row to Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: TreeNode, v: int, d: int) -> TreeNode:
        if d == 1:
            newR = TreeNode(v, root)
            return newR
        
        queue = collections.deque([root])
        for _ in range(d - 2):
            for _ in range(len(queue)):
                node = queue.popleft()
                if node.left: queue.append(node.left)
                if node.right: queue.append(node.right)
        
        while queue:
            node = queue.popleft()
            temp = node.left
            node.left = TreeNode(v)
            node.left.left = temp
            temp = node.right
            node.right = TreeNode(v)
            node.right.right = temp
        
        return root

File: Breadth-First Search/Binary tree/LET_no623_Add One Row to Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        if (depth == 1)
            return new TreeNode(val, root, nullptr);

        queue<TreeNode*> q;
        q.push(root);

        while (depth-- > 2) {
            int n = q.size();
            while (n-- > 0) {
                TreeNode* cur = q.front();
                q.pop();
                if (cur->left) 
                    q.push(cur->left);
                if (cur->right) 
                    q.push(cur->right);
            }
        }

        TreeNode *cur;
        while (!q.empty()) {
            cur = q.front();
            q.pop();
            cur->left = new TreeNode(val, cur->left, nullptr);
            cur->right = new TreeNode(val, nullptr, cur->right);
        }

        return root;
    }
};